packages/compose/src/hooks/use-async-list/index.ts

```diff
@@ -65,7 +65,10 @@ function useAsyncList< T >(
 				...state,
 				...list.slice( nextIndex, nextIndex + step ),
 			] );
-			asyncQueue.add( {}, append( nextIndex + step ) );
+
+			queueMicrotask( () =>
+				asyncQueue.add( {}, append( nextIndex + step ) )
+			);
```

Comment:
This may have an unexpected side effect on other areas of the editor that use the `useAsyncList` hook. Have you measured any potential effects there?

I wonder if alternatively, we can just debounce the showing of the pattern previews instead of the addition of patterns to the async list?

Reply:
> I wonder if alternatively, we can just debounce the showing of the pattern previews instead of the addition of patterns to the async list?

That's the whole purpose of "useAsyncList" debounce the showing of the previews. But instead of relying on random "debounce" value, it relies on `requestIdleCallback` internally (in priority queue) to continue rendering previews as long as the browser is idle.

Reply:
Makes sense to me. 

Do you have any concerns with shipping this change @youknowriad? I wasn't able to measure any noticeable difference before and after this PR in my testing.

Reply:
The change is fine but I'm surprised it's needed. Maybe @fullofcaffeine know the difference here

Reply:
> The change is fine but I'm surprised it's needed. Maybe @fullofcaffeine know the difference here

The difference for me is significant to the point of crashing the browser sometimes. You can see the comparison in the screencasts above. I don't know why my system is more susceptible to this issue though, but I'm concerned that if it affected my system - which is quite powerful - it could affect a good portion of users, too.

Reply:
> We have a bug that's fixed by changing a subtle detail deep inside the core and there's no comment. Seems like a very good place to say "We cannot use asyncQueue.add() here because …" wink

Ah, good point :) I'll add a comment and link back to the issue, too.

Reply:
not that I have deep knowledge about the callback, but the big change is that when calling `asyncQueue.add( {}, append( nextIndex + step ) )` we're calling that _in context_ immediately after calling `setCurrent()`, while when calling it in a micro-task it guarantees that any other micro-tasks enqueued before this call will run.

my guess is that React issues its own micro-task on that `setCurrent` call and the ordering here matters. swap the order of the `setCurrent()` call and the `queueMicrotask()` call and I would bet you'd recreate the bug.

remember that until a function _returns_ all synchronous code executes before anything else happens, then all micro-tasks are run in the order they were queued, and then all tasks are run in the order they were queued.

the effective change here is a (probably) re-ordering of micro-tasks, otherwise it wouldn't make sense to me either how this changes things.

Reply:
@dmsnell That makes sense, thanks for sharing.

Reply:
Ok I now understand what's happening here :) and yeah the explanation above is related:

 So with `useAsyncList` what we want to delay is the actual rendering of the list one by one and between each "rendering" we want to check if there's enough time in the browser's idle time to perform another rendering. The way we do that is to use a priority queue with the list of callback. so we do this for each callback:
 
 ```
 asyncQueue.add( {}, append( nextIndex + step ) )
 ```
 
 The difference here is that the callback in the queue does different things between React 17 and React 18:
 
  - React 17: append sets state **inside** the callback, and setting state causes the component to render **synchronously**, in other words, the rendering is happening within the priority queue callback, so once the callback is finished (callback returned), priority queue checks whether there's remaining time to continue rendering or let the browser perform higher priority tasks.
  - React 18: append sets state **inside** the callback but setting state doesn't cause a synchronous render, instead the callback schedules a micro-task to render but returns to the caller. In this case, the caller is the priority queue and update the state (update an array) is very quick meaning, we'll have enough time to "fill" the array entirely before even the micro tasks to render the first component runs, this results in all the components for the lazy list being rendered all at the same time.

In other words `useAsyncList` is doing nothing in React 18.

---
Solutions: 

The current solution proposed in this PR only appends the queue after React renders meaning the priority queue will be able to check between two renders whether there's enough time to continue its cutie. There's a subtle difference with the previous solution in the sense that previously we were able to "load" the queue with multiple "appends" and let the priority queue render as time was available. 

So it's a fine fix and we can ship it but it makes me think that the priority queue is kind of useless now. So it's like we're doing something like:

```
const append = ( nextIndex: number ) => () => {
	if ( list.length <= nextIndex ) {
		return;
	}
	setCurrent( ( state ) => [
		...state,
		...list.slice( nextIndex, nextIndex + step ),
	] );
	
	window.requestIdleCallback( () => { 
	    asyncQueue.add( {}, append( nextIndex + step ) );
       });
};
```

Basically, with this PR we'll be rendering a single item (or a **step** number of items) each 40ms (approximative requestIdleCallback minimum wait time). We're not using the "window" offered by requestIdleCallback to render as much iterations as possible, instead of only run 1 iteration. So basically, the async list fills slower than it **can** load. This may have some effects in performant browser like chromes where you may notice that the patterns in the inserter show up slower than expected (where potentially it was almost instant).

The idea solution is to add the callbacks synchronously but make sure the rendering is done "synchronously" within the callbacks. I'm not sure we have a way in React 18 to perform a synchronous re-rendering though. Maybe @jsnajdr knows. 

In the meantime, I'm ok with this temporary fix.

Reply:
> I'm not sure we have a way in React 18 to perform a synchronous re-rendering though.

There is `ReactDOM.flushSync( () => setState( ... ) )` which will do the updates in the callback synchronously, which in React 18 means "in the next promise microtask." There are no truly sync updates in React 18.

Other places where updates are sync (in the microtask sense) are store updates in `useSyncExternalStore`, and also class component's `this.forceUpdate()`. That's all AFAIK.

Reply:
> ReactDOM.flushSync( () => setState( ... ) )

Oh great, meaning, we should be able to update the code here to always "add" all the iterations and wrap the setState call with flushSync. 

Something close to (pseudo-code):

```
for ( i = 0; i < list.length; I += step ) {
queue.add( {}, () => {
    ReactDOM.flushSync( () => setCurrent( ( state ) => [
		...state,
		...list.slice( i, i + step ),
    ] ) );
} );
```

Reply:
```js
class PriorityQueue {
	add( queue, callback ) {
		callback();
	}
}
```

🤡

Reply:
When originally adapting `useAsyncList` for concurrent mode, I didn't notice that we're checking `timeRemaining()` and trying to fit as much as possible work into one idle callback. I thought it's just a series of idle callbacks each performing `step` renders. Exactly the behavior we're getting in this branch after the `queueMicrotask`.

Here's how we could get the original behavior even in concurrent mode:
```js
async function renderStep( nextIndex ) {
  // schedules setState, including the rendering, for next microtask tick
  ReactDOM.flushSync( () => {
    setCurrent( state => state.concat( list.slice( nextIndex, nextIndex + step ) ) );
  } );
  // schedules a microtask that finishes _after_ the render is done
  await Promise.resolve();

  return nextIndex + step;
}

async function runWaitingList( deadline ) {
  while ( nextIndex < list.length ) {
    // add new items to the list, render, and wait until the rendering finishes
    nextIndex = await renderStep( nextIndex );

    // if we still have time, continue rendering in this idle callback
    if ( deadline.timeRemaining() > 0 ) {
      continue;
    }

    // otherwise schedule a new idle callback, with a new deadline, and
    // continue rendering after it fires.
    deadline = await new Promise( ( resolve ) => {
      requestIdleCallback( ( d ) => resolve( d ) );
    } );
  }
}

requestIdleCallback( runWaitingList );
```

The magical trick here is that the idle callback can be an async function, doing its work asynchronously, with promises, and still keep checking the `deadline`!. It doesn't need to be fully sync.

The code I wrote doesn't use the `priority-queue`, it calls `requestIdleCallback` directly.

To make it work with `priority-queue`, we'd have to:
1. Make the `runWaitingList` function async, and make it `await callback()`.
2. Add all the work to the queue at once, we don't need the tail recursion for `asyncQueue.add` that we have now. We know upfront what we want to get done, and then it will get executed in idle callbacks.

```js
for ( let i = 0; i < list.length; i += step ) {
  const nextList = list.slice( 0, i + step );
  asyncQueue.add( {}, async () => {
    ReactDOM.flushSync( () => setCurrent( nextList ) );
    await Promise.resolve();
  } );
}
```

One thing about `flushSync` is that it's in the `react-dom` package. For mobile, we'd have to use the `react-native` equivalent, which I hope exists. And export them from `@wordpress/element`.

By the way, I think that in concurrent mode we eventually shouldn't need `useAsyncList` at all. Processing a long render queue asynchronously, taking smaller bites, checking deadlines, and letting higher-priority interruptions run, that's what concurrent mode is supposed to do natively. I don't have much insight into the precise details right now, but once we learn about that, we should be able to eliminate `useAsyncList`.

Reply:
Ok so `ReactDOM.flushSync` doesn't re-render synchronously? So in that case we still need the microtask. In that case, yeah the only part forward would be to promisify priority-queue (or built a promisified alternative). Also, I don't see why we need `flushSync` in this case? what does it do? `setCurrent` already schedules the microtask internally

Reply:
> By the way, I think that in concurrent mode we eventually shouldn't need useAsyncList at all. Processing a long render queue asynchronously, taking smaller bites, checking deadlines, and letting higher-priority interruptions run, that's what concurrent mode is supposed to do natively. I don't have much insight into the precise details right now, but once we learn about that, we should be able to eliminate useAsyncList.

Yeah, I agree but it seems React is y et to provide the APIs to define the priorities and basically `useAsyncList` could probably but one of these APIs.

Reply:
> Ok so ReactDOM.flushSync doesn't re-render synchronously?

Turns out I was wrong abou this: `flushSync` is indeed completely synchronous, including effects. That means we don't need any microtasks and promises, only a `flushSync` wrapper around `setCurrent`. I submitted this fix as #48238.

> Also, I don't see why we need `flushSync` in this case? what does it do? `setCurrent` already schedules the microtask internally

I'm not 100% sure if the scheduling of `setState` is always the microtask. The scheduling strategy is different in different scenarios, and sometimes the update can be scheduled for later.